### DISCUSSION TITLE
C# Core Guideline

### DISCUSSION THREAD

**Discussion by mavasani** [dotnet/csharplang#1779] [Language Ideas] [answered]
Ported from https://github.com/dotnet/roslyn-analyzers/issues/1731. Filed by @magol 

-----------------------------

For the C++ world, it is a [C++ Core Guideline](http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines). For some of the guidelines it is analysis in Visual Studio, but many of the guidelines are more of a philosophical nature.

I can not find any similar resource for C# that have the same level of detail and big scope. I suggest to create a GIT repository to start collect suggestion from the community and build a **C# Core Guideline**.

----------------------------

**Comment by svick** [comment, marked as answer]
C++ needs such guidelines, because, mostly due to historical reasons, it has many parts that are unsafe and hard to use correctly. C# is in a much better position, since it has much less issues of that kind. And when there is something that people commonly get wrong, there's usually an article in the official documentation that explains the right approach (keep in mind that C++ doesn't have any "official documentation", only the standard).

So, how would such Core Guidelines improve upon the current situation? And even if there was a need for something like that, shouldn't it be part of [the documentation](https://docs.microsoft.com/en-us/dotnet/index)?

There's also [Framework Design Guidelines](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/), which deal specifically with designing the public API of a library.
